modules/dev-utils/src/main/java/org/apache/ignite/development/utils/IgniteIndexReader.java

```diff
@@ -0,0 +1,2130 @@
+        File file =  new File(cacheWorkDir, partId == INDEX_PARTITION ? INDEX_FILE_NAME : String.format(PART_FILE_TEMPLATE, partId));
+
+        if (!file.exists())
+            return null;
+        else if (partId == -1)
+            print("Analyzing file: " + file.getPath());
+
+        return file;
+    }
+
+    /** */
+    private void readPage(FilePageStore store, long pageId, ByteBuffer buf) throws IgniteCheckedException {
+        try {
+            store.read(pageId, buf, false);
+        }
+        catch (IgniteDataIntegrityViolationException e) {
+            // Replacing exception due to security reasons, as IgniteDataIntegrityViolationException prints page content.
+            throw new IgniteException("Failed to read page, id=" + pageId + ", file=" + store.getFileAbsolutePath());
+        }
+    }
+
+    /**
+     * Read index file.
+     */
+    public void readIdx() {
+        long partPageStoresNum = Arrays.stream(partStores)
+            .filter(Objects::nonNull)
+            .count();
+
+        print("Partitions files num: " + partPageStoresNum);
+
+        Map<Class, AtomicLong> pageClasses = new HashMap<>();
+
+        long pagesNum = idxStore == null ? 0 : (idxFile.length() - idxStore.headerSize()) / pageSize;
+
+        print("Going to check " + pagesNum + " pages.");
+
+        Map<Class, Set<Long>> pageIoIds = new HashMap<>();
+
+        AtomicReference<Map<String, TreeTraversalInfo>> treeInfo = new AtomicReference<>();
+
+        AtomicReference<Map<String, TreeTraversalInfo>> horizontalScans = new AtomicReference<>();
+
+        AtomicReference<PageListsInfo> pageListsInfo = new AtomicReference<>();
+
+        List<Throwable> errors;
+
+        ProgressPrinter progressPrinter = new ProgressPrinter(System.out, "Reading pages sequentially", pagesNum);
+
+        try {
+            errors = scanFileStore(INDEX_PARTITION, FLAG_IDX, idxStore, (pageId, addr, io) -> {
+                progressPrinter.printProgress();
+
+                pageClasses.computeIfAbsent(io.getClass(), k -> new AtomicLong(0)).incrementAndGet();
+
+                if (io instanceof PageMetaIO) {
+                    PageMetaIO pageMetaIO = (PageMetaIO)io;
+
+                    long metaTreeRootId = normalizePageId(pageMetaIO.getTreeRoot(addr));
+
+                    treeInfo.set(traverseAllTrees("Index trees traversal", metaTreeRootId, CountOnlyStorage::new, this::traverseTree));
+
+                    treeInfo.get().forEach((name, info) -> {
+                        info.innerPageIds.forEach(id -> {
+                            Class cls = name.equals(META_TREE_NAME)
+                                ? IndexStorageImpl.MetaStoreInnerIO.class
+                                : H2ExtrasInnerIO.class;
+
+                            pageIoIds.computeIfAbsent(cls, k -> new HashSet<>()).add(id);
+                        });
+
+                        pageIoIds.computeIfAbsent(BPlusMetaIO.class, k -> new HashSet<>()).add(info.rootPageId);
```

Comment:
Is info.rootPageId differ from metaTreeRootId?

Reply:
Yes, metaTreeRootId is root for meta tree, and info.rootPageId is root of each index tree.